### PR TITLE
allow zig js host functions to return JSError

### PR DESCRIPTION
**Pull request by nektro**
progress towards zig JSValue stability and safety in https://github.com/oven-sh/bun/issues/159
more followups to come

### CODE REVIEW

**Comment by robobun** [comment]
@nektro, your commit 4b7dafc2637b957bd165d412a407901bcd7a6ada is building: [`#5947`](https://buildkite.com/bun/bun/builds/5947)

**Review by paperclover — COMMENTED** [review]
should we actually be annotating every function with `bun.JSError!`? i think my solution to this in dave/more-bake is better, since it doesnt include the `catch` control flow on functions that don't return an error

**Comment by nektro** [comment]
nearly every function annotated will be taking advantage of it eventually and unless there's a noticeable perf effect then the consistency makes the function's purpose very clear at a glance

**Comment by paperdave** [comment]
if we want all of these functions to have the explicit annotation, it should be `bun.OOM || bun.JSError`, which i have in my branch defined as `bun.JSOOM` (can be named something else). the wrapper checks and handles both of these errors, if specified.

this is why i asked to wait a day for my branch to finish, as i'm nearly close to merging it. was this blocking something else?

**Comment by nektro** [comment]
in a couple more stages i think it would make more sense for `OutOfMemory` to be a part of `bun.JSError`
pr is only blocking itself

**Review by paperclover — APPROVED** [review]

File: src/bun.zig [hunk 1]
```diff
@@ -91,20 +91,11 @@ pub inline fn namedAllocator(comptime name: [:0]const u8) std.mem.Allocator {
...
-    ///
-    /// Prefer `global.jsErrorFromCPP()` over constructing this error manually,
-    /// that way the exception is asserted to actually exist.
-    ///
-    /// To pass this to C++ code as JSC::JSValue, use `global.errorUnionToCPP(...)`
+    /// There is an active exception on the global object.
+    /// You should almost never have to construct this manually.
     JSError,
+    // XXX: This is temporary! meghan will remove this soon
+    OutOfMemory,
```
Comment (dylan-conway) [inline src/bun.zig:98]: I think we should consider keeping this here and rename `error.JSError` to `error.ThrownException` or something similar

Comment (nektro) [reply src/bun.zig:98]: yeah there's many followups ahead of this

**Review by dylan-conway — APPROVED** [review]

**Pull request merged**